File: scripts/coupons_get_token.py
import json
import boto3
import bcrypt
import jwt
from datetime import datetime, timedelta

# Constants
SECRET = "0HXc4w5NEzA61HkV"
DYNAMODB_TABLE = "users"

def handler(event, context):
    try:
        body = json.loads(event.get("body", "{}"))
        username = body.get("username")
        password = body.get("password")

        if not username or not password:
            return {
                "statusCode": 403,
                "body": json.dumps({"message": "Invalid credentials"})
            }

        # Retrieve user data from DynamoDB
        dynamodb = boto3.resource("dynamodb", endpoint_url="http://localhost:4566")
        table = dynamodb.Table(DYNAMODB_TABLE)
        response = table.get_item(Key={"username": username})
        user = response.get("Item")

        if not user or not bcrypt.checkpw(password.encode(), user["password"].encode()):
            return {
                "statusCode": 403,
                "body": json.dumps({"message": "Invalid credentials"})
            }

        # Generate JWT token
        expiration_time = datetime.now(datetime.timezone.utc) + timedelta(hours=1)
        token = jwt.encode(
            {"username": username, "exp": expiration_time},
            SECRET,
            algorithm="HS256"
        )

        return {
            "statusCode": 200,
            "body": json.dumps({
                "token": token,
                "expiresAt": expiration_time.isoformat() + "Z"
            })
        }
    except Exception as e:
        print(f"Error: {e}")
        return {
            "statusCode": 500,
            "body": json.dumps({"message": "Internal server error"})
        }

File: scripts/coupons_get_token.js
const AWS = require('aws-sdk');
const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');

const dynamo = new AWS.DynamoDB.DocumentClient();
const SECRET = '0HXc4w5NEzA61HkV';
const TOKEN_EXPIRATION = '1h';

exports.handler = async (event) => {
    try {
        const { username, password } = JSON.parse(event.body);

        if (!username || !password) {
            return { statusCode: 403, body: JSON.stringify({ message: 'Invalid credentials' }) };
        }

        // Query the users table
        const params = {
            TableName: 'users',
            Key: { username },
        };

        const user = await dynamo.get(params).promise();

        if (!user.Item || !(await bcrypt.compare(password, user.Item.password))) {
            return { statusCode: 403, body: JSON.stringify({ message: 'Invalid credentials' }) };
        }

        // Generate JWT
        const token = jwt.sign({ username }, SECRET, { expiresIn: TOKEN_EXPIRATION });
        return {
            statusCode: 200,
            body: JSON.stringify({
                token,
                expiresAt: new Date(Date.now() + 60 * 60 * 1000).toISOString(),
            }),
        };
    } catch (error) {
        console.error(error);
        return { statusCode: 500, body: JSON.stringify({ message: 'Internal server error' }) };
    }
};
